controller/grpc.go

```diff
@@ -0,0 +1,1014 @@
+
+func (g *grpcStatsHandler) TagRPC(ctx context.Context, rpcTagInfo *stats.RPCTagInfo) context.Context {
+	return ctx
+}
+
+func (g *grpcStatsHandler) HandleRPC(context.Context, stats.RPCStats) {
+}
+
+const ctxKeyRemoteHost = "remote-host"
+
+func (g *grpcStatsHandler) TagConn(ctx context.Context, connTagInfo *stats.ConnTagInfo) context.Context {
+	remoteHost, _, _ := net.SplitHostPort(connTagInfo.RemoteAddr.String())
+	ctx = context.WithValue(ctx, ctxKeyRemoteHost, remoteHost)
+	return ctx
+}
+
+func (g *grpcStatsHandler) HandleConn(context.Context, stats.ConnStats) {
+}
+
+func (g *grpcAPI) streamInterceptor(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) (err error) {
+	ctx, logRequestEnd := g.logRequest(stream.Context(), info.FullMethod)
+	defer func() {
+		logRequestEnd(ctx, err)
+	}()
+
+	ctx, err = g.authorize(stream.Context())
+	if err != nil {
+		return err
+	}
+
+	wrappedStream := middleware.WrapServerStream(stream)
+	wrappedStream.WrappedContext = ctx
+	return handler(srv, wrappedStream)
+}
+
+func (g *grpcAPI) unaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (res interface{}, err error) {
+	var logRequestEnd func(context.Context, error)
+	ctx, logRequestEnd = g.logRequest(ctx, info.FullMethod)
+	defer func() {
+		logRequestEnd(ctx, err)
+	}()
+
+	ctx, err = g.authorize(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return handler(ctx, req)
+}
+
+func (g *grpcAPI) Status(context.Context, *empty.Empty) (*protobuf.StatusResponse, error) {
+	healthy := true
+	if err := g.db.Exec("ping"); err != nil {
+		healthy = false
+	}
+	return protobuf.NewStatusResponse(healthy, nil), nil
+}
+
+func (g *grpcAPI) listApps(req *protobuf.StreamAppsRequest) ([]*protobuf.App, *data.PageToken, error) {
+	pageSize := int(req.GetPageSize())
+	pageToken, err := data.ParsePageToken(req.PageToken)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if pageSize > 0 {
+		pageToken.Size = pageSize
+	} else {
+		pageSize = pageToken.Size
+	}
+
+	appIDs := protobuf.ParseIDsFromNameFilters(req.GetNameFilters(), "apps")
+	ctApps, nextPageToken, err := g.appRepo.ListPage(data.ListAppOptions{
+		PageToken:    *pageToken,
+		AppIDs:       appIDs,
+		LabelFilters: protobuf.NewControllerLabelFilters(req.GetLabelFilters()),
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if pageSize == 0 {
+		pageSize = len(ctApps)
+	}
+
+	if len(appIDs) == 0 {
+		appIDs = nil
+	}
+
+	apps := make([]*protobuf.App, 0, pageSize)
+	n := 0
+	for _, a := range ctApps {
+		apps = append(apps, protobuf.NewApp(a))
+		n++
+
+		if n == pageSize {
+			break
+		}
+	}
+
+	return apps, nextPageToken, nil
+}
+
+func (g *grpcAPI) StreamApps(req *protobuf.StreamAppsRequest, stream protobuf.Controller_StreamAppsServer) error {
+	unary := !(req.StreamUpdates || req.StreamCreates)
+
+	var apps []*protobuf.App
+	var nextPageToken *data.PageToken
+	var appsMtx sync.RWMutex
```

Comment:
Is this mutex needed?

Reply:
Seems it can be dropped. I think a previous iteration of this method was a two-way stream which had multiple go routines interacting with `apps`.